### ISSUE TITLE
Links displayed in "Your Privacy" measurements section, no longer have links underneath on dev server

### ISSUE THREAD

**Issue by SoftVision-CosminMuntean** [mozilla/testpilot#2103]
[Notes]:
- This seems to be a regression introduced by #2048.

[Affected versions]:
- Firefox 51.0.1

[Affected Platforms]:
- All Windows
- All Mac
- All Linux

[Steps to reproduce]:
1. Open Firefox with a new profile and navigate to "https://testpilot.dev.mozaws.net.com" page.
2. Click on any available experiment card (except Min Vid and Tab Center). 
3. Scroll down to the page and observe the "Your Privacy" measurements section.
4. Click on the available links from the gray background section.

[Expected result]:
- You are redirected to the specific pages.

[Actual result]:
- The text links have no URL underneath and clicking them will only select the string.

[Additional Notes]:
- Here is a screen recording with the issue: https://goo.gl/UqWAEi.

**Comment by jaredhirsch** [comment]
Confirmed this bug is legit, talked with @stasm about it a bit.

Basically, the `href` attribute is getting stripped out of the anchor tag when the translation is inserted. This is a deliberate l20n design decision that prevents localizers from inserting evil links in place of the intended href. So, the suggestion from stas was to move the anchor tag contents into the source code:

```<li data-l10n-id={this.l10nId(['measurements', idx])}><a href="https://www.mozilla.org/privacy/websites/" target="_blank"></a></li>```

and out of the translation:

```
pageshotMeasurements3 = For all visits to the Page Shot website, the data we collect is 
described in our <a>Websites, Communications & Cookies Privacy Notice</a>.
```

This `href` replacement concept is discussed more in the [l20n docs](https://github.com/l20n/spec/blob/master/dom-overlays.markdown).

Unfortunately, since we generate all the measurements lines inside a [loop](https://github.com/mozilla/testpilot/blob/master/frontend/src/app/containers/ExperimentPage.js#L396-L397):

```js
<ul>
{measurements.map((note, idx) => (
  <li data-l10n-id={this.l10nId(['measurements', idx])}>{note}</li>
))}
</ul>
```

the link localization approach built into l20n won't work for us here.

A code fix seems like it will inevitably be hacky/gross:

- detect when l20n is done, either by polling the DOM, or upgrading l20n (stas offered to release a new version tomorrow that exposes a `document.l10n.ready` promise that would resolve when it's done)

- after l20n is done, grab the experiment's strings, loop over the privacy/measurements section of the DOM, and if the strings contain any `href` attributes, shove those back into the corresponding anchor tag

I'm usually all for ugly hacks that work, but this offends even my sense of code smell.

The best alternative is probably to remove links from the measurements section altogether, but then we're invalidating strings again. We'd also need to find a way for experiments to share links to stuff.

I guess, as a compromise, we could come up with a convention like: replace hrefs with slugs (like `<a href='page-shot-privacy'>Privacy</a>`) in the strings, then put the corresponding URLs somewhere else in the yaml file. Although then we still wind up doing a third pass at the DOM, after React builds it and after L20n inserts strings.

In short, there doesn't seem to be a simple, clean way around this bug. Anyone else have suggestions?

**Comment by jaredhirsch** [comment]
I'm thinking React could look at the strings and pre-template the anchor tag into the DOM, so that l20n "just works" and we don't need a third pass. that's likely the cleanest of the unclean approaches

**Comment by ghost** [comment]
We can allow unsafe HTML if we need an emergency fix, right?

**Comment by jaredhirsch** [comment]
> We can allow unsafe HTML if we need an emergency fix, right?

Sadly, no. I asked about just this sort of thing, but @stasm said that's not available in l20n right now

**Comment by stasm** [comment]
Hey, sorry about the trouble. I'll take a deeper look at this tomorrow morning and see if there's anything else that we can do to fix this on l20n's side.

**Comment by ghost** [comment]
I didn't use the right words.  I'm talking about dangerouslySetInnerHTML() which we were originally using on the strings: https://github.com/mozilla/testpilot/pull/2048/files#diff-d863820a9512305ab17dd7dd0768ea4bL388

**Comment by stasm** [comment]
> stas offered to release a new version tomorrow that exposes a document.l10n.ready promise that would resolve when it's done

I tagged version `4.0.0-beta.1` which includes the `document.l10n.ready` promise which is resolved when the localization of the whole `document` finishes.

I have another idea for a possible solution. Looking at the following code it occurs to me that the URL should be part of the `measurement` data structure, just like the `l10n-id` would if it wasn't built dynamically.

```jsx
<ul>
{measurements.map((note, idx) => (
  <li data-l10n-id={this.l10nId(['measurements', idx])}>{note}</li>
))}
</ul>
```

I wasn't able to find what `measurements` look like, but could something like the following be possible?

```jsx
{measurements.map((measure, idx) => (
  measure.url ?
    <li data-l10n-id={this.l10nId(['measurements', idx])}>
      <a href={measure.url}></a>
    </li>
    :
    <li data-l10n-id={this.l10nId(['measurements', idx])}></li>
))}
```

**Comment by jaredhirsch** [comment]
Unassigning myself. Don't think I'll get to this today, and I'm out next week

**Comment by SoftVision-CosminMuntean** [comment]
The "Embedly" link from Activity Stream experiment page and "Archive.org", "here" links from NoMore404s experiment page are still not working. 

The rest of the links from privacy section are working as expected.

**Comment by SoftVision-CosminMuntean** [comment]
@6a68 The "Embedly" link from Activity Stream experiment page is still broken on dev, stage and production servers.
Or the fix will land only on stage and production after deploy?

**Comment by dannycoates** [comment]
@SoftVision-CosminMuntean the embedly link won't be fixed this sprint

**Comment by pdehaan** [comment]
Semi-related, but we should probably check w/ the **activity-stream** team, as I believe they dropped Embedly usage (in favor of **fathom**) somewhat recently.
